kvserver: account for AddSSTableRequests in QPS

kvoli opened the pull request:
Previously, Queries-Per-Second (QPS) was calculated uniformly per
`BatchRequest` as 1. This patch introduces variable QPS calculation
for `AddSSTableRequest`, which use an order of magnitude more
resources than other request types.

This patch introduces the
`kv.replica_stats.addsst_request_size_factor` cluster setting. This
setting is used to attribute QPS to `AddSSTableRequest` sizes. The
calculation is done as QPS = 1 + size(AddSSTableRequest) / factor.
When `kv.replica_stats.addsst_request_size_factor` is less than 1, or
no `AddSSTableRequest` exists within a `BatchRequest`, then QPS = 1;
the current behavior today.

resolves #73731

Release note (performance improvement):
Introduced `kv.replica_stats.addsst_request_size_factor` cluster
setting. This setting is used to tune Queries-Per-Second (QPS)
sensitivity to large imports. By default, this setting is disabled.
When enabled, the size of any AddSSTableRequest will contribute to
QPS in inverse relation to this settings magnitude. By default this
setting configured to a conservative 50,000; every 50 kilobytes will
be accounted for as an additional 1 QPS.

kvoli committed:
kvserver: account for AddSSTableRequests in QPS

Previously, Queries-Per-Second (QPS) was calculated uniformly per
`BatchRequest` as 1. This patch introduces variable QPS calculation
for `AddSSTableRequest`, which use an order of magnitude more
resources than other request types.

This patch introduces the
`kv.replica_stats.addsst_request_size_factor` cluster setting. This
setting is used to attribute QPS to `AddSSTableRequest` sizes. The
calculation is done as QPS = 1 + size(AddSSTableRequest) / factor.
When `kv.replica_stats.addsst_request_size_factor` is less than 1, or
no `AddSSTableRequest` exists within a `BatchRequest`, then QPS = 1;
the current behavior today.

resolves #73731

Release note (performance improvement):
Introduced `kv.replica_stats.addsst_request_size_factor` cluster
setting. This setting is used to tune Queries-Per-Second (QPS)
sensitivity to large imports. By default, this setting is disabled.
When enabled, the size of any AddSSTableRequest will contribute to
QPS in inverse relation to this settings magnitude. By default this
setting configured to a conservative 50,000; every 50 kilobytes will
be accounted for as an additional 1 QPS.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76252)
<!-- Reviewable:end -->

kvoli committed:
kvserver: update replica_send call flow docs

This patch updates the function naming to reflect the current method
handler for batch requests containing at least one write request. to
`executeWriteBatch`, from `executeReadWriteBatch`.

Release note: None

aayushshah15 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76252)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[pkg/kv/kvserver/replica_metrics.go, line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-Mvdj3eq6SVjHs8B9hvp:-Mvdj3eq6SVjHs8B9hvq:bxbug60) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_metrics.go#L253)):*
> ```Go
> // that weren't sent through a DistSender, which in practice should be
> // practically none). Also return the amount of time over which the stat was
> // accumulated. AddSSTableRequests can be treated differently, in which case
> ```

nit: This seems like an odd place to have this comment, could we just have it closer to where we're actually accounting for QPS and have this comment point to that location?
___
*[pkg/kv/kvserver/replica_send.go, line 974 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdjSlG4CJ3igutGTrd:-MvdjSlG4CJ3igutGTre:bujtsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_send.go#L974)):*
> ```Go
> // 1 QPS, unless an AddSSTableRequest exists, in which case the sum of all
> // AddSSTableRequest's data size is divided by a factor and added to QPS. This
> // specific treatment of QPS is a special case to accout for the difference
> ```

account*
___
*[pkg/kv/kvserver/replica_send.go, line 975 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdjWHJBipQ1dsfdeLi:-MvdjWHJBipQ1dsfdeLj:bt6i0fr) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_send.go#L975)):*
> ```Go
> // AddSSTableRequest's data size is divided by a factor and added to QPS. This
> // specific treatment of QPS is a special case to accout for the difference
> // extreme between AddSSTableRequest and other requests in terms of resource
> ```

extreme difference? 
___
*[pkg/kv/kvserver/replica_stats.go, line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdkdFP2VxQTaMk5fr4:-MvdkdFP2VxQTaMk5fr5:b-w4ivg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_stats.go#L46)):*
> ```Go
> 	"kv.replica_stats.addsst_request_size_factor",
> 	"the divisor that is applied to addsstable request sizes, then recorded in a leaseholders QPS; 0 means all requests are treated as cost 1",
> 	0,
> ```

Did we decide that this factor was going to be zero by default? That's kind of unfortunate. Could we instead consider setting it to a large-ish value (say 1mb)? 
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 86 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdnjEk66xYpTRutpdp:-MvdnjEk66xYpTRutpdq:b-j6rd9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_rankings_test.go#L86)):*
> ```Go
> // TestAddSSTQPSStat verifies that AddSSTableRequests are accounted for
> // differently, when present in a BatchRequest, with a divisor set.
> func TestAddSSTQPSStat(t *testing.T) {
> ```

nice test!
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdnHnr4QuPUhEZWaTI:-MvdnHnr4QuPUhEZWaTJ:b-uafmuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_rankings_test.go#L92)):*
> ```Go
> 
> 	tc := serverutils.StartNewTestCluster(t, 1, base.TestClusterArgs{
> 		ServerArgs: base.TestServerArgs{
> ```

I'd also recommend disabling the `replicateQueue` by using `ReplicationMode: Manual` here to avoid lease transfers away from the replica you're measuring the QPS off of. 
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-Mvdl7op5aZFrq3mChnX:-Mvdl7op5aZFrq3mChnY:bif7ewj) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_rankings_test.go#L93)):*
> ```Go
> 	tc := serverutils.StartNewTestCluster(t, 1, base.TestClusterArgs{
> 		ServerArgs: base.TestServerArgs{
> 			DisableWebSessionAuthentication: true,
> ```

I'm surprised we need this. Let's remove it if we don't?
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdmE6p3T_iDni70lKU:-MvdmE6p3T_iDni70lKV:b-5j5fny) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_rankings_test.go#L167)):*
> ```Go
> 
> 		// If queries are correctly recorded, we should see increase in query count by
> 		// the expected QPS. However, it is possible to to get a slightly higher or
> ```

`to to` 

also nit: the wrapping is a little off here. 


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76252)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @kvoli)

___
*[pkg/kv/kvserver/replica_metrics.go, line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-Mvdj3eq6SVjHs8B9hvp:-MveOYrf74_FJxfvyDfQ:b-82opcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_metrics.go#L253)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

nit: This seems like an odd place to have this comment, could we just have it closer to where we're actually accounting for QPS and have this comment point to that location?
</blockquote></details>

Removed. Where this is being accounted for, it is recorded - were you think of adding a comment above the method call `r.getBatchRequestQPS()`? Currently this is an abridged version of the function comment `getBatchRequest`.
___
*[pkg/kv/kvserver/replica_send.go, line 974 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdjSlG4CJ3igutGTrd:-MvePCnG-TBSpu_lhSkI:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_send.go#L974)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

account*
</blockquote></details>

Fixed.
___
*[pkg/kv/kvserver/replica_send.go, line 975 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdjWHJBipQ1dsfdeLi:-MvePDZGATB6A58vGSzJ:b-qj92pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_send.go#L975)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

extreme difference? 
</blockquote></details>

Reworded.
___
*[pkg/kv/kvserver/replica_stats.go, line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdkdFP2VxQTaMk5fr4:-MveRvwiA4RqjKIVwsON:b-vh3vjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_stats.go#L46)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Did we decide that this factor was going to be zero by default? That's kind of unfortunate. Could we instead consider setting it to a large-ish value (say 1mb)? 
</blockquote></details>

Updated following our discussion to a conservative 50k.
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 86 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdnjEk66xYpTRutpdp:-MveRRl68C2SiNqU8Gfz:b32wy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_rankings_test.go#L86)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

nice test!
</blockquote></details>

ty
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdnHnr4QuPUhEZWaTI:-MveRTI49PJ8vuiHSP_r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_rankings_test.go#L92)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

I'd also recommend disabling the `replicateQueue` by using `ReplicationMode: Manual` here to avoid lease transfers away from the replica you're measuring the QPS off of. 
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-Mvdl7op5aZFrq3mChnX:-MveRU1o1awmGSHdjDnD:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_rankings_test.go#L93)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

I'm surprised we need this. Let's remove it if we don't?
</blockquote></details>

Removed.
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdmE6p3T_iDni70lKU:-MveS-Fr4OdU35hkEORq:besgv6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_rankings_test.go#L167)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

`to to` 

also nit: the wrapping is a little off here. 
</blockquote></details>

Formatted.


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-:-MveVA7OBAjRQgYwDJo3:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @kvoli)

___
*[pkg/kv/kvserver/replica_metrics.go, line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-Mvdj3eq6SVjHs8B9hvp:-MveTylj63gxlFLfBop_:b-xu9pct) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_metrics.go#L253)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Removed. Where this is being accounted for, it is recorded - were you think of adding a comment above the method call `r.getBatchRequestQPS()`? Currently this is an abridged version of the function comment `getBatchRequest`.
</blockquote></details>

Yeah, this looks good to me. I was thinking that we'd have a line here saying "See getBatchRequestQPS() for how this accounted for". Hopefully, this makes it such that when the QPS accounting method is changed, these comments point to a single place that has an up-to-date comment.


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76252)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @kvoli)

___
*[pkg/kv/kvserver/replica_stats.go, line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvdkdFP2VxQTaMk5fr4:-MveVSTc4F8Er_4x9nby:b-5ya5l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/931838327ac308b8c2e051ad97eb4d60f801a113/pkg/kv/kvserver/replica_stats.go#L46)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Updated following our discussion to a conservative 50k.
</blockquote></details>

Let's update the release note with that.


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
Added Nathan in case he wants to give this a pass as well.

kvoli commented on pkg/kv/kvserver/replica_rankings_test.go:
```diff
@@ -73,3 +80,91 @@ func TestReplicaRankings(t *testing.T) {
+	// Construct an sst with 100 keys that will be reused with different divisors.
+	sstKeys := make([]sstutil.KV, 100)
+	for i := range sstKeys {
+		sstKeys[i] = sstutil.KV{fmt.Sprintf("%3d", i+1), 1, "value"}
```
Use scratch key, then concat this as a suffix.

kvoli commented on pkg/kv/kvserver/replica_rankings_test.go:
```diff
@@ -73,3 +80,91 @@ func TestReplicaRankings(t *testing.T) {
+		// QPS. However, it is possible to to get a slightly higher or lower number, due to rounding
+		// or timing errors; the result is calculated using floating point division. We allow a
+		// tolerance of 4 to avoid flaking the test due to this inaccuracy.
+		require.InDelta(t, queriesIncrease, testCase.expectedQPS, 4)
```
Expose a method testGetQPSSum to assert exactly on the QPS recorded.

kvoli replied:
Exposed method, however here it doesn't seem possible to avoid the issue with inaccuracy. Investigating the cause further.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 11 files at r5, 11 of 11 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-:-MvzQQ-dEsZ9hfDpyoDS:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15 and @kvoli)

___
*[pkg/kv/kvserver/replica_stats.go, line 53 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvzPZFPDwqp9lqdV-N-:-MvzPZFPDwqp9lqdV-N0:b1zfdmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a304e38cecfcc0e2b4c652aa47c4eb8b1bc3b17/pkg/kv/kvserver/replica_stats.go#L53)):*
> ```Go
> ).WithPublic()
> 
> // AddSSTableRequestSizeFactor returns the size cost factor for a given replica.
> ```

nit: I think this method actually hurts readability because it forces readers to jump over here to confirm that this is pulling from a cluster setting and it's not a per-range property. We generally don't wrap cluster setting accesses like this in methods. What do you think about inlining it?


<!-- Sent from Reviewable.io -->

kvoli commented:
bors r+

kvoli commented:
bors r-

kvoli commented:
> <img alt=":lgtm:" width="61" height="20" src="https://camo.githubusercontent.com/341bcbc4fc105a8a54332ced9ba9f1163b5fd81578777fffe7568934ec76d9d7/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67">
> 
> > Reviewed 11 of 11 files at r5, 11 of 11 files at r6, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-:-MvzQQ-dEsZ9hfDpyoDS:bnfp4nl)_ status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15 and @kvoli)
> 
> _[pkg/kv/kvserver/replica_stats.go, line 53 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76252#-MvzPZFPDwqp9lqdV-N-:-MvzPZFPDwqp9lqdV-N0:b1zfdmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a304e38cecfcc0e2b4c652aa47c4eb8b1bc3b17/pkg/kv/kvserver/replica_stats.go#L53)):_
> 
> > ```go
> > ).WithPublic()
> > 
> > // AddSSTableRequestSizeFactor returns the size cost factor for a given replica.
> > ```
> 
> nit: I think this method actually hurts readability because it forces readers to jump over here to confirm that this is pulling from a cluster setting and it's not a per-range property. We generally don't wrap cluster setting accesses like this in methods. What do you think about inlining it?

Updated to be inline.

irfansharif commented:
The commit message refers to https://github.com/cockroachdb/cockroach/pull/7371, I think we meant https://github.com/cockroachdb/cockroach/issues/73731?

kvoli commented:
> The commit message refers to #7371, I think we meant #73731?

Good spot. Updated.

kvoli commented:
bors r+
